src/sdk.js

```diff
@@ -95,24 +96,55 @@ class OpenForm {
     CSPNonce.setValue(CSPNonceValue);
     initialiseSentry(sentryDSN, sentryEnv);
 
-    // ensure that the basename has no trailing slash (for react router)
-    let pathname = basePath || window.location.pathname;
-    if (pathname.endsWith('/')) {
-      pathname = pathname.slice(0, pathname.length - 1);
+    let pathname;
+    if (!this.useHashRouting) {
+      pathname = basePath || window.location.pathname;
+      if (pathname.endsWith('/')) {
+        // ensure that the basename has no trailing slash (for react router)
+        pathname = pathname.slice(0, pathname.length - 1);
+      }
+    } else {
+      pathname = '';
     }
     this.basePath = pathname;
+    this.makeRedirect();
     this.calculateClientBaseUrl();
   }
 
+  makeRedirect() {
+    // Perform pre-redirect based on this action: this is decoupled from the backend
+    const query = new URLSearchParams(document.location.search);
+    const action = query.get('_of_action');
+    if (action) {
+      // as actionParams is in the form `q1%3Dquery1%26q2%3Dquery2`, we first URL decode and then parse it:
+      const actionParams = new URLSearchParams(decodeURIComponent(query.get('_of_action_params')));
+      query.delete('_of_action');
+      query.delete('_of_action_params');
+
+      const redirectPath = getRedirectPathname(action, actionParams);
+
+      let newUrl;
+      if (!this.useHashRouting) {
+        newUrl = `${window.location.origin}${window.location.pathname}/${redirectPath}${
+          query.size ? '?' + query : ''
+        }`;
+      } else {
+        // No pathname with hash based routing
+        // TODO use query.size once we have better browser support
+        newUrl = `${window.location.origin}/${
+          [...query].length ? '?' + query : ''
+        }#/${redirectPath}`;
```

Comment:
can we perhaps avoid the string concatenation and instead use the URL class?

```
const url = new URL(location);
url.searchParams.set("foo", "bar");
history.pushState({}, "", url);
```
from MDN: https://developer.mozilla.org/en-US/docs/Web/API/History/pushState#change_a_query_parameter

I was trying to figure out if we need to URL encode params for `pushState` or not and my question isn't really answered yet, but it seems that using the `new URL(...)` approach allows us to rely on the browser for the correct behaviour.